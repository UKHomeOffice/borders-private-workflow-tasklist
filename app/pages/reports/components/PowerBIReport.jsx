import React, { Component, Fragment } from 'react';
import { service, factories, models } from 'powerbi-client';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import axios from 'axios';
import secureLocalStorage from '../../../common/security/SecureLocalStorage';
import LogoBar from '../../../core/components/LogoBar';

export class PowerBIReport extends Component {
  constructor(props) {
    super(props);
    const { useMobileLayout } = this.props;
    this.reportContainer = React.createRef();
    this.setFullscreen = this.setFullscreen.bind(this);
    this.onPageChange = this.onPageChange.bind(this);
    this.useMobileLayout = useMobileLayout;
    this.visitedPages = [];
  }

  componentDidMount() {
    const { apiRefUrl, accessToken, embedUrl, id, token } = this.props;

    const powerBIBranchNames = [
      'Central',
      'Detection Services',
      'Heathrow',
      'Intelligence',
      'National Operations',
      'North',
      'South',
      'South East and Europe',
    ];

    const {
      team: { branchid: branchId } = {},
    } = secureLocalStorage.get('shift');

    axios({
      method: 'get',
      url: `${apiRefUrl}/v2/entities/branch?filter=id=eq.${branchId}`,
      headers: { Authorization: `Bearer ${token}` },
    })
      .then(({ data }) => {
        const branchName = data.data.length && data.data[0].name;
        if (powerBIBranchNames.includes(branchName)) {
          this.userBranchName = branchName;
        }
        const embedConfig = {
          accessToken,
          type: 'report',
          embedUrl,
          id,
          permissions: models.Permissions.Read,
          settings: {
            filterPaneEnabled: false,
            layoutType: this.useMobileLayout
              ? models.LayoutType.MobilePortrait
              : models.LayoutType.Master,
          },
          tokenType: models.TokenType.Embed,
        };

        const powerbi = new service.Service(
          factories.hpmFactory,
          factories.wpmpFactory,
          factories.routerFactory,
        );
        const target = this.reportContainer.current;
        this.report = target && powerbi.embed(target, embedConfig);

        if (this.userBranchName) {
          this.report && this.report.on('pageChanged', this.onPageChange);
        }
      })
      .catch(error => {
        console.log(`Error: ${error}`);
      });
  }

  onPageChange(e) {
    const {
      newPage,
      newPage: { displayName },
    } = e.detail;

    if (this.visitedPages.includes(displayName)) return;

    let visualNumber;
    let target;
    this.visitedPages.push(displayName);
    const { userBranchName } = this;
    if (displayName === 'Command Brief - OAR') {
      visualNumber = 5;
      target = { table: 'OAR', column: 'Branch Name' };
    } else if (displayName === 'Command Brief - IEN') {
      visualNumber = 4;
      target = { table: 'IEN', column: 'Branch' };
    } else {
      return;
    }
    newPage.getVisuals().then(visuals => {
      const targetVisual = visuals[visualNumber];
      targetVisual
        .setSlicerState({
          filters: [
            {
              $schema: 'http://powerbi.com/product/schema#basic',
              target,
              operator: 'In',
              values: [userBranchName],
            },
          ],
        })
        .catch(errors => {
          console.log(`Power BI could not get visuals. Errors: ${errors}`);
        });
    });
  }

  setFullscreen() {
    if (this.report) this.report.fullscreen();
  }

  render() {
    const logoBar = this.useMobileLayout ? null : (
      <LogoBar setFullscreen={this.setFullscreen} />
    );

    return (
      <Fragment>
        <div
          id="report"
          style={{ width: '100%', height: '100%' }}
          ref={this.reportContainer}
        />
        {logoBar}
      </Fragment>
    );
  }
}

PowerBIReport.propTypes = {
  accessToken: PropTypes.string.isRequired,
  apiRefUrl: PropTypes.string.isRequired,
  embedUrl: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  useMobileLayout: PropTypes.bool.isRequired,
};

export default connect(state => {
  return {
    token: state.keycloak.token,
    apiRefUrl: state.appConfig.apiRefUrl,
  };
})(PowerBIReport);
