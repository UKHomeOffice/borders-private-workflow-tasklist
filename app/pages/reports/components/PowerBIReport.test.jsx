import React from 'react';
import PowerBIReport from './PowerBIReport';

describe('PowerBIReport', () => {
  const props = {
    accessToken: 'xxx',
    embedUrl: 'http://www.example.com',
    id: 'abc',
    name: 'Power BI Report',
    useMobileLayout: true,
  };

  it('renders a PowerBIReport', () => {
    const wrapper = shallow(<PowerBIReport {...props} />);
    expect(wrapper.find('Report').exists()).toEqual(true);
  });

  it('does not render a LogoBar if narrow screen', () => {
    const wrapper = shallow(<PowerBIReport {...props} />);
    expect(wrapper.find('LogoBar').exists()).toEqual(false);
  });

  it('renders a LogoBar if wide screen', () => {
    const wrapper = shallow(
      <PowerBIReport {...{ ...props, useMobileLayout: false }} />,
    );
    expect(wrapper.find('LogoBar').exists()).toEqual(true);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<PowerBIReport {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
