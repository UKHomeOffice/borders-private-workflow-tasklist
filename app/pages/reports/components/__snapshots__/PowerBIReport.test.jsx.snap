// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PowerBIReport matches snapshot 1`] = `
<Report
  accessToken="xxx"
  embedId="abc"
  embedType="report"
  embedUrl="http://www.example.com"
  pageName="Power BI Report"
  permissions="Read"
  reportMode="view"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  tokenType="Embed"
/>
`;
